types/scene/subsurface_tree.c

```diff
@@ -0,0 +1,215 @@
+	struct wlr_scene_subsurface_tree *surface_tree =
+		wl_container_of(addon, surface_tree, surface_addon);
+	return surface_tree;
+}
+
+static void surface_tree_reconfigure(
+		struct wlr_scene_subsurface_tree *surface_tree) {
+	struct wlr_surface *surface = surface_tree->surface;
+
+	struct wlr_scene_node *prev = NULL;
+	struct wlr_subsurface *subsurface;
+	wl_list_for_each(subsurface, &surface->current.subsurfaces_below,
+			current.link) {
+		struct wlr_scene_subsurface_tree *child =
+			surface_tree_from_subsurface(surface_tree, subsurface);
+		if (prev != NULL) {
+			wlr_scene_node_place_above(&child->tree->node, prev);
+		}
+		prev = &child->tree->node;
+
+		wlr_scene_node_set_position(&child->tree->node,
+			subsurface->current.x, subsurface->current.y);
+	}
+
+	if (prev != NULL) {
+		wlr_scene_node_place_above(&surface_tree->scene_surface->node, prev);
+	}
+	prev = &surface_tree->scene_surface->node;
+
+	wl_list_for_each(subsurface, &surface->current.subsurfaces_above,
+			current.link) {
+		struct wlr_scene_subsurface_tree *child =
+			surface_tree_from_subsurface(surface_tree, subsurface);
+		wlr_scene_node_place_above(&child->tree->node, prev);
+		prev = &child->tree->node;
+
+		wlr_scene_node_set_position(&child->tree->node,
+			subsurface->current.x, subsurface->current.y);
+	}
+}
+
+static void surface_tree_handle_surface_commit(struct wl_listener *listener,
+		void *data) {
+	struct wlr_scene_subsurface_tree *surface_tree =
+		wl_container_of(listener, surface_tree, surface_commit);
+
+	// TODO: only do this on subsurface order or position change
+	surface_tree_reconfigure(surface_tree);
+}
+
+static void surface_tree_addon_destroy(struct wlr_addon *addon) {
+	struct wlr_scene_subsurface_tree *surface_tree =
+		wl_container_of(addon, surface_tree, surface_addon);
+	wlr_scene_node_destroy(&surface_tree->tree->node);
+}
+
+static const struct wlr_addon_interface surface_tree_addon = {
+	.name = "wlr_scene_subsurface_tree",
+	.destroy = surface_tree_addon_destroy,
+};
+
+static struct wlr_scene_subsurface_tree *scene_surface_tree_create(
+	struct wlr_scene_node *parent, struct wlr_surface *surface);
+
+static bool surface_tree_create_subsurface(
+		struct wlr_scene_subsurface_tree *surface_tree,
+		struct wlr_subsurface *subsurface) {
+	struct wlr_scene_subsurface_tree *subsurface_tree = scene_surface_tree_create(
+		&surface_tree->tree->node, subsurface->surface);
+	if (surface_tree == NULL) {
+		return false;
+	}
+
+	subsurface_tree->parent = surface_tree;
+	wlr_addon_init(&subsurface_tree->surface_addon,
+		&subsurface->surface->addons, surface_tree, &surface_tree_addon);
+
+	return true;
+}
+
+static void surface_tree_handle_surface_new_subsurface(
+		struct wl_listener *listener, void *data) {
+	struct wlr_scene_subsurface_tree *surface_tree =
+		wl_container_of(listener, surface_tree, surface_new_subsurface);
+	struct wlr_subsurface *subsurface = data;
+	surface_tree_create_subsurface(surface_tree, subsurface);
```

Comment:
This is the one remaining place I see where allocation failure isn't handled.

Reply:
Indeed, added a `wl_resource_post_no_memory` call.